src/query/functions/aggregation/function.go

```diff
@@ -122,24 +122,39 @@ func stddevFn(values []float64, bucket []int) float64 {
 }
 
 func varianceFn(values []float64, bucket []int) float64 {
-	sum, count := sumAndCount(values, bucket)
-
-	// Cannot take population standard deviation of less than 1 value
-	if count < 1 {
+	if len(values) == 0 || len(bucket) == 0 {
 		return math.NaN()
 	}
 
-	average := sum / count
-	sumOfSquares := 0.0
+	// Using Welford's online algorithm for calculating variance
+	// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
+	//
+	// This algorithm is used in Prometheus and also should provide better numerical precision than
+	// the straight-forward implementation of the variance formula. The algorithm iterates through the values
+	// and at the each step recalculates mean and variance of the values seen so far.
+
+	count := 0
+	partialMean := 0.0
+	partialVarTimesCount := 0.0 // for better precision, calculate `variance * count` and divide at the end
```

Comment:
Nit: consider putting these into a `var (...)` block. Usually it looks cleaner that way.

Reply:
https://github.com/m3db/m3/pull/2799/commits/e5a9d981953cde0c2a051bd129a4506dea1f15dc